src/main/java/org/zalando/nakadi/service/EventPublisher.java

```diff
@@ -99,6 +99,7 @@ EventPublishResult publishInternal(final String events,
 
             validate(batch, eventType);
             partition(batch, eventType);
+            compact(batch, eventType);
```

Comment:
There is an alternative solution. When building the effective Schema, you can make this field mandatory and leverage the feature of validation, so that it doesn't come this point. I know it somehow couples validation with compaction, but I guess that's already the case for almost all aspects of event publishing.

More specifically, extending https://github.com/zalando/nakadi/blob/167c0c1206b83128f5ec16b5808d54ec20c75c82/src/main/java/org/zalando/nakadi/validation/JsonSchemaEnrichment.java#L36 by adding another `if` and then modifying the schema by marking `partition_compaction_key` to be `required` and to be longer than 0 character. The nice thing about doing that is that it provides an error with the complete path to the missing field.

Then the error would also be reported as a problem during validation, which I think makes more sense than "partitioning".

Reply:
Yes, makes sense. Changed https://github.com/zalando/nakadi/pull/907/commits/7ab73d2309358778495a30321d864422a1bc4a68

Reply:
we introduced another loop here, do you think it is time to combine it and do every step in one loop?

Reply:
@adyach in the API we always point out which step failed. So if something fails during erichment, for example, it means that everything is fine regarding validation. So if we did as you suggest, it could happen that for the first event it validates fine but then enrichment breaks. The user fix the payload, send again and this time if fails validation for the next event. So I think that even though it would be possible to do what you suggest, I think it would be more work then it looks like to not change the current behaviour.